ESQL: Introduce per agg filter

costin opened the pull request:
Add support for aggregation scoped filters that work dynamically on the
 data in each group.

```
| STATS
    success = COUNT(*) WHERE 200 <= code AND code < 300,
   redirect = COUNT(*) WHERE 300 <= code AND code < 400,
 client_err = COUNT(*) WHERE 400 <= code AND code < 500,
 server_err = COUNT(*) WHERE 500 <= code AND code < 600,
total_count = COUNT(*)
```

Implementation wise, the base AggregateFunction has been extended to
 allow a filter to be passed on. This is required to incorporate the
 filter as part of the aggregate equality/identify which would fail with
 the filter as an external component.

As part of the process, the serialization for the existing aggregations
 had to be fixed so AggregateFunction implementations so that it
 delegates to their parent first.

costin committed:
ESQL: Introduce per agg filter

Add support for aggregation scoped filters that work dynamically on the
 data in each group.

| STATS
    success = COUNT(*) WHERE 200 <= code AND code < 300,
   redirect = COUNT(*) WHERE 300 <= code AND code < 400,
 client_err = COUNT(*) WHERE 400 <= code AND code < 500,
 server_err = COUNT(*) WHERE 500 <= code AND code < 600,
 total_count = COUNT(*)

Implementation wise, the base AggregateFunction has been extended to
 allow a filter to be passed on. This is required to incorporate the
 filter as part of the aggregate equality/identify which would fail with
 the filter as an external component.

As part of the process, the serialization for the existing aggregations
 had to be fixed so AggregateFunction implementations so that it
 delegates to their parent first.

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4:
```diff
@@ -119,17 +119,15 @@ fields
-    | qualifiedName ASSIGN booleanExpression
+    : (qualifiedName ASSIGN)? booleanExpression
     ;
 
 fromCommand
     : FROM indexPattern (COMMA indexPattern)* metadata?
     ;
 
 indexPattern
-    : clusterString COLON indexString
-    | indexString
+    : (clusterString COLON)? indexString
```
Small improvements to the grammar that don't change the parsing.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -301,6 +302,11 @@ private static void checkInvalidNamedExpressionUsage(
         Set<Failure> failures,
         int level
     ) {
+        // unwrap filtered expression
+        if (e instanceof FilteredExpression fe) {
+            e = fe.delegate();
+            // TODO add verification for filter clause
+        }
```
Still TBD

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/AggregateFunction.java:
```diff
@@ -52,25 +54,51 @@ public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
 
     protected AggregateFunction(StreamInput in) throws IOException {
-        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class));
+        this(
+            Source.readFrom((PlanStreamInput) in),
+            in.readNamedWriteable(Expression.class),
+            in.getTransportVersion().onOrAfter(TransportVersions.ESQL_PER_AGGREGATE_FILTER)
+                ? in.readNamedWriteable(Expression.class)
+                : Literal.TRUE,
+            in.getTransportVersion().onOrAfter(TransportVersions.ESQL_PER_AGGREGATE_FILTER)
+                ? in.readNamedWriteableCollectionAsList(Expression.class)
+                : emptyList()
+        );
```
Uniform serialization across all aggregate functions - we should had that since the beginning to avoid having each subclass add its own repetitive serialization.
/cc @nik9000

nik9000 replied:
I think it's easier to make the filter wrap the aggregation rather than inside each one. I think also that serializing a list in case aggs use it is probably not a great choice either. It's fine, but feels like something we'd do when we need it.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Avg.java:
```diff
@@ -74,19 +79,26 @@ public DataType dataType() {
     protected NodeInfo<Avg> info() {
-        return NodeInfo.create(this, Avg::new, field());
+        return NodeInfo.create(this, Avg::new, field(), filter());
     }
 
     @Override
     public Avg replaceChildren(List<Expression> newChildren) {
-        return new Avg(source(), newChildren.get(0));
+        return new Avg(source(), newChildren.get(0), newChildren.get(1));
+    }
+
+    @Override
+    public Avg withFilter(Expression filter) {
+        return new Avg(source(), field(), filter);
```
The side effect is all aggs had to be modified to take into account the parent filter.

nik9000 replied:
Could we wrap them with something like we do at runtime? It feels like that might work.

costin replied:
Not sure what you mean. My initial approach was to wrap the agg and filter in separate class (FilteredAggregate with two children - aggregate and filter/expression ) however this messed up the optimizer since the rules picked the aggregate functions but had no idea there was a filter associated with it.
This is problematic for scenario where the same agg is defined with different filter:
`STATS c = count(). c_f = count() WHERE a > 1`
since the where filter is disjoined from the count, the rules replaced the second count with the first one but that is incorrect since they have different execution paths.
Furthermore at compute time the filter and actual agg gets fused anyway so the two are a pair anyway.

nik9000 replied:
They do get fused anyway, but it's just a lot more convenient if every agg doesn't have to worry about being filtered. If `filter` is a member here we have to think about it all over the place. It's enough that the fused filter at compute time has to add all the tests for it, I'd love to avoid that bubbling all the way out. OTOH, if that's how it has to be, I'm not deep enough to object.

fang-xing-esql replied:
From performance perspective, one of the advantages of separating filter from aggregation is that, if the same filter applies to different aggregations, it makes it easier to recognize that it is the same filter and can be evaluated once for all aggregations that share it, in stead of evaluating each filter within each aggregation separately. However I couldn’t figure out how the current infrastructure supports it yet. At first look at this, it reminds me of the (common)table expressions and union all subqueries in SQL, but we don’t have this infrastructure in ES|QL yet, and it could be a much broader scope.

costin replied:
This approach doesn't prevent optimizing the filter evaluation.
Each aggregation ends up having its own filter because `count(*)` is different than `count(*) where a > 10`. Initially they were decoupled however the rules considered them the same since the filter was sitting outside the aggregation.
When dealing with the same filter, we can assemble a different pipeline so the mask or masked block is reused across multiple aggregations.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceStatsAggExpressionWithEval.java:
```diff
@@ -69,18 +66,18 @@ protected LogicalPlan rule(Aggregate aggregate) {
         int[] counter = new int[] { 0 };
 
-        for (NamedExpression agg : aggs) {
+        for (int i = 0, s = aggs.size(); i < s; i++) {
+            NamedExpression agg = aggs.get(i);
+
```
This needs to be rolled back.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java:
```diff
@@ -251,19 +257,22 @@ private void aggregatesToFactory(
                                 );
                             }
                         } else {
-                            sourceAttr = aggregateFunction.inputExpressions().stream().map(e -> {
-                                Attribute attr = Expressions.attribute(e);
+                            // extra dependencies like TS ones (that require a timestamp)
+                            for (Expression input : aggregateFunction.references()) {
+                                Attribute attr = Expressions.attribute(input);
```
replace inputExpressions() with the existing references().
That method could be replaced in a separate PR or later on.

costin replied:
I've removed this now - references does what's needed (looks at the dependencies and removes constants) tagging @dnhatn fyi.

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
```diff
@@ -2290,3 +2290,60 @@ from employees
+        min = min(salary), min_a = min(salary) where salary > 99999  
+;
+
+max:integer |max_a:integer|min:integer | min_a:integer
+74999       |null         |25324       | null        
+;
```
This will fail since the filter is incorrectly applies (internally seen is marked as true as the implementation expects a constant to skip a whole block but this isn't one - just one big block of `false`).

costin replied:
/cc @nik9000

nik9000 replied:
https://github.com/elastic/elasticsearch/pull/113804 should cover it.

costin replied:
It does.

costin reviewed (commented):
Draft PR to check the CI.
Was based on an old commit to avoid some of the volatility in main.
Will rebase and polish after a few CI rounds.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

elasticsearchmachine commented:
Pinging @elastic/kibana-esql (ES|QL-ui)

costin reviewed (commented):
Addressed TBDs, add verifier code thus promoted from draft and rebased on latest main.

costin committed:
Update docs/changelog/113735.yaml

elasticsearchmachine commented:
Hi @costin, I've created a changelog YAML for you. *Note* that since this PR is labelled `release highlight`, you need to update the changelog YAML to fill out the extended information sections.

costin committed:
Fix spotless

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/FilteredExpression.java:
```diff
@@ -0,0 +1,93 @@
+import static java.util.Arrays.asList;
+
+/**
+ * Basic wrapper for expressions declared with a nested filter (typically in stats).
```
Maybe let's add that this is used only during parsing and shouldn't exist after function resolution?

alex-spies reviewed (commented):
I had a first round, looking good so far. I'll give this another look tomorrow.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/AggregateFunction.java:
```diff
@@ -52,25 +54,51 @@ public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
+    }
+
+    @Deprecated(since = "8.16", forRemoval = true)
+    protected void deprecatedWriteTo(StreamOutput out) throws IOException {
```
Suggestion: rename to `deprecatedWriteParameters` to be more precise.

alex-spies reviewed (changes requested):
Another round.

There's a serialization bug in ToPartial that should be addressed before merging.

Additionally, filters are not respected by the aggs on constants that still rely on `COUNT(*)`; currently I think that's SUM and COUNT. This should be addressed or we should open an issue for this.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -122,6 +123,7 @@ protected static Batch<LogicalPlan> substitutions() {
             "Substitutions",
             Limiter.ONCE,
             new ReplaceLookupWithJoin(),
+            new SubstituteFilteredExpression(),
```
I think it's important to comment why this is needed: this has to run before `ReplaceStatsNestedExpressionWithEval` and `ReplaceStatsAggExpressionWithEval` can be applied.

alex-spies replied:
@costin maybe this was overlooked due to having a thumbs-up from Bogdan already :D A small comment would be great - either here or on `SubstituteFilteredExpression`.

alex-spies reviewed (commented):
I think we're being inconsistent with the aggregate names:
```
row a = 1 | stats sum(a) where a > 3, sum(a)

    sum(a)     
---------------
1      
```
To avoid surprising behavior, I think this should return 2 columns: `sum(a) where a > 3` and `sum(a)`.

Ordering this example's aggs the other way around is particularly surprising:
```
row a = 1 | stats sum(a), sum(a) where a > 3

    sum(a)     
---------------
null    
```
This looks like the `sum(a)` (without filter) returns `null`, but that's not true.

alex-spies reviewed (commented):
Ok, all done now!

Found a couple more things to validate and test.

bpintea commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
```diff
@@ -2290,3 +2290,60 @@ from employees
+from employees
+| stats max = max(salary), max_f = max(salary) where salary < 50000,
+        min = min(salary), min_f = min(salary) where salary > 50000,
+        exp_gt = max(salary) - min(salary) where salary > 50000,
+        exp_lt = max(salary) - min(salary) where salary < 50000
```
Also tests which:
* filter on other fields/expressions_with_fields that are agg'ed on;
* group, groups.

bpintea replied:
`FROM employees | STATS max(height) WHERE gender IN ("M") BY gender, is_rehired` trips an assertion:
```aiignore
[ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [runTask-0] fatal error in thread [elasticsearch[runTask-0][esql_worker][T#11]], exiting java.lang.AssertionError: java.lang.IllegalStateException
: can't release already released object [BooleanArrayVector[positions=100, values=[true, false, true, true, true, false, false, true, false, false, ...]]]
        at org.elasticsearch.base@9.0.0-SNAPSHOT/org.elasticsearch.core.Releasables.closeExpectNoException(Releasables.java:49)
        at org.elasticsearch.compute.operator.HashAggregationOperator$1AddInput.close(HashAggregationOperator.java:156)
        at org.elasticsearch.compute.operator.HashAggregationOperator.addInput(HashAggregationOperator.java:169)
        at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:266)
        at org.elasticsearch.compute.operator.Driver.run(Driver.java:189)
        at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:378)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:34)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:992)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalStateException: can't release already released object [BooleanArrayVector[positions=100, values=[true, false, true, true, true, false, false, true, false, false, ...]]]
        at org.elasticsearch.compute.data.AbstractNonThreadSafeRefCounted.decRef(AbstractNonThreadSafeRefCounted.java:40)
        at org.elasticsearch.compute.data.AbstractNonThreadSafeRefCounted.close(AbstractNonThreadSafeRefCounted.java:59)
        at org.elasticsearch.base@9.0.0-SNAPSHOT/org.elasticsearch.core.IOUtils.close(IOUtils.java:71)
        at org.elasticsearch.base@9.0.0-SNAPSHOT/org.elasticsearch.core.IOUtils.close(IOUtils.java:87)
```

bpintea commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
> we don't propagate the filter into the `COUNT(*)` in `Sum.surrogate`

There might be more to it, since 
`from test | stats count(*) where a > 3` also fails (and same `from test | stats count(a) where a > 3`), but `from test | stats c_gt3 = count(*) where a > 3, c_lt3 = count(*) where a < 3` works (and that's how the last added test is actually correct).

costin replied:
That's because there's no true folding of the aggregation but instead we try to rewrite them into MV_functions + case.
Which fails short when having to evaluate the filter (which might be foldable or not).

costin replied:
As a side note, when the filters are foldable, they evaluate to either true (essentially discarded) or false meaning the agg won't run and can be folded to its initial value, 0 for count and `null` for the other aggs.

alex-spies replied:
Alright, let's address the const case later.

bpintea replied:
Opened https://github.com/elastic/elasticsearch/issues/115522 to call this out explicitly as it's actually producing incorrect results.

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
I'm not sure I understand correctly; but
```
| STATS s = sum(field) WHERE field > 0 AND s > 0
```
is invalid because the `WHERE` inside the aggregation refers to the result of the aggregation. To be correct, the second part of the predicate needs to be moved into a separate `WHERE` command:
```
| STATS s = sum(field) WHERE field > 0 | WHERE s > 0
```

costin committed:
Merge branch 'main' into esql/per-agg-filtering

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java:
```diff
@@ -274,16 +283,28 @@ private void aggregatesToFactory(
+                    aggSupplier = supplier(aggregateFunction, inputChannels);
+
+                    // apply the filter only in the initial phase - as the rest of the data is already filtered
+                    if (aggregateFunction.hasFilter() && mode.isInputPartial() == false) {
```
This would make sense only if the filter is common across all aggregations so it can filter the data coming in across all aggs - see https://github.com/elastic/elasticsearch/issues/114352

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Avg.java:
```diff
@@ -74,19 +79,26 @@ public DataType dataType() {
 
-        return field().foldable() ? new MvAvg(s, field) : new Div(s, new Sum(s, field), new Count(s, field), dataType());
+        return field().foldable()
+            ? new MvAvg(s, field)
```
See https://github.com/elastic/elasticsearch/issues/100634#issuecomment-2400665066

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/FilteredExpression.java:
```diff
@@ -0,0 +1,93 @@
+/**
+ * Basic wrapper for expressions declared with a nested filter (typically in stats).
+ */
+public class FilteredExpression extends Expression implements SurrogateExpression {
```
I've removed it for now. This is exactly the type of scenario, SurrogateExpression was added yet because it is used to fold aggregations, it now fails to serve its initial purpose.

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
Good catch however I believe this is a bug. Grouping functions should only be allowed in the BY clause - here it could work it's the same as a grouping however I tried it and we allow a bucket with different field and argument.
/cc @bpintea

bpintea replied:
> I tried it and we allow a bucket with different field and argument

The semantic equality for `BUCKET`s is currently run over the aggs in `STATS` against the `GroupingFunction`s in its `BY` clause. But since `checkInvalidNamedExpressionUsage()` isn't run on all `AggregateFunction` children and since the filter-on-agg is now added within the `AggregateFunction` itself, the condition isn't caught.
So this the check needs to be extended. But IMO it should either be a follow-up of this PR, or part of it, I don't think it's a general issue.

bpintea replied:
> I believe this is a bug
This still works and didn't find an existing issue, so opened https://github.com/elastic/elasticsearch/issues/115521

costin committed:
Address feedback
Add more integration tests
Fix nullable filter expanding to the agg

costin committed:
Merge branch 'main' into esql/per-agg-filtering

costin commented:
Thanks for the feedback folks - especially @alex-spies . I've addressed most of the issues except:
- improving the verifier
- didn't add any handling regarding aggregation folding over constants when dealing with a filter; see my comment. Looking at the code I think this needs a different approach however in meantime we can prevent queries where a filter is specified through the verifier.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java:
```diff
@@ -672,6 +677,33 @@ public List<Alias> visitFields(EsqlBaseParser.FieldsContext ctx) {
+        Source source = source(ctx);
+        Alias field = visitField(ctx.field(), source);
+        var filterExpression = ctx.booleanExpression();
+        Expression condition = filterExpression != null ? expression(filterExpression) : null;
+
+        if (condition != null) {
```
```suggestion

        if (filterExpression != null) {
            Expression condition = expression(filterExpression);
```

expression() will throw, rather than return `null`.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java:
```diff
@@ -274,16 +283,28 @@ private void aggregatesToFactory(
-                    } else {
-                        throw new EsqlIllegalArgumentException("aggregate functions must extend ToAggregator");
+
+                    AggregatorFunctionSupplier aggSupplier = null;
+                    aggSupplier = supplier(aggregateFunction, inputChannels);
```
```suggestion
                    AggregatorFunctionSupplier aggSupplier = supplier(aggregateFunction, inputChannels);
```

costin committed:
Merge branch 'main' into esql/per-agg-filtering

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java:
```diff
@@ -672,6 +677,33 @@ public List<Alias> visitFields(EsqlBaseParser.FieldsContext ctx) {
+        Source source = source(ctx);
+        Alias field = visitField(ctx.field(), source);
+        var filterExpression = ctx.booleanExpression();
+        Expression condition = filterExpression != null ? expression(filterExpression) : null;
+
+        if (condition != null) {
```
Not following. expression is called only if the context is not null so it should return null.

bpintea replied:
It's just a nit: `filterExpression` and `condition` have the same nullability, since `expression()` will never return `null`: if  `filterExpression` is not-null, `condition` will be not-null. So no need to check twice against `null`.

costin replied:
ahhhh - now I see it!
Thanks for bearing with me.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java:
```diff
@@ -672,6 +677,33 @@ public List<Alias> visitFields(EsqlBaseParser.FieldsContext ctx) {
+            // allow condition only per aggregated function
+            else {
+                source = source(filterExpression);
+                throw new ParsingException(source, "WHERE clause allowed only for aggregate functions [{}]", source.text());
```
I've updated this so the aggregate function and the filter is returned with the error pointing to the filter location.
That's because a STATS can be quite large if multiple aggregations are specified.

costin committed:
Improve verifier + add more tests

costin commented:
Fixed the verifier and addressed the feedback.

alex-spies reviewed (approved):
I still need to fully re-review this tomorrow, but I see the serialization bug in ToPartial is gone so please consider this unblocked!

costin committed:
Address nit

costin committed:
Merge branch 'main' into esql/per-agg-filtering

costin committed:
Fix incorrect file merge

costin committed:
Polish

costin committed:
Fix hasFilter method
Disable source check for Rate function

costin committed:
Improve Rate constructor

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -537,6 +537,24 @@ public void testCombineProjectionWithDuplicateAggregation() {
+    public void testStatsWithFilteringDefaultAliasing() {
+        var plan = plan("""
+            from test
+            | stats sum(salary), sum(salary) WheRe last_name ==   "Doe"
```
Nice, finally some proper use of sarcastic case.

alex-spies reviewed (approved):
Alright, all done now.

Nice! IMO this is in a very nice state now and LGTM.

So far, I still believe that aggs on consts with filters _should_ be do-able with the current approach - but maybe I'm not thinking of the right counter examples. In any case, I agree that we should do that in a follow-up PR (after revisiting the overall approach).

costin committed:
Merge remote-tracking branch 'remotes/upstream/main' into esql/per-agg-filtering

costin committed:
Polish

costin committed:
Merge branch 'main' into esql/per-agg-filtering

costin committed:
Improve FoldNull

costin committed:
Merge branch 'main' into esql/per-agg-filtering

costin committed:
More polish

costin committed:
Merge branch 'main' into esql/per-agg-filtering

costin committed:
Spotless

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/EsqlFunctionRegistry.java:
```diff
@@ -259,19 +259,21 @@ private FunctionDefinition[][] functions() {
-                def(Top.class, Top::new, "top"),
-                def(Values.class, Values::new, "values"),
-                def(WeightedAvg.class, WeightedAvg::new, "weighted_avg") },
+                def(Avg.class, uni(Avg::new), "avg"),
```
This was needed to help the compiler disambiguate which constructor to pick as there are multiple variants.

costin committed:
Fix compiler warning

costin committed:
Another set of fixes

costin committed:
Merge branch 'main' into esql/per-agg-filtering

elasticsearchmachine commented:
## 💔 Backport failed
The backport operation could not be completed due to the following error:
```
An unexpected error occurred when attempting to backport this PR.
```

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 113735`

bpintea commented:
From [above](https://github.com/elastic/elasticsearch/pull/113735#issuecomment-2401302220):

> we can prevent queries where a filter is specified through the verifier.

@costin, I guess this hasn't been implemented right? I've opened https://github.com/elastic/elasticsearch/issues/115522, but do we want a fix before addressing that?
